tools/undelete-blocks/main.go

```diff
@@ -0,0 +1,431 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/grafana/dskit/flagext"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/storage/tsdb/block"
+	"github.com/grafana/mimir/pkg/util/objtools"
+)
+
+type config struct {
+	bucketConfig   objtools.BucketConfig
+	includeTenants flagext.StringSliceCSV
+	excludeTenants flagext.StringSliceCSV
+	inputJSONFile  string
+	dryRun         bool
+}
+
+func (c *config) registerFlags(f *flag.FlagSet) {
+	c.bucketConfig.RegisterFlags(f)
+	f.Var(&c.includeTenants, "include-tenants", "If not empty, only blocks for these tenants are considered.")
+	f.Var(&c.excludeTenants, "exclude-tenants", "Blocks for these tenants are not considered.")
+	f.StringVar(&c.inputJSONFile, "input-json-file", "", "If set, the blocks to undelete are read from the provided json file.")
+	f.BoolVar(&c.dryRun, "dry-run", false, "If true, no writes/deletes are performed and are instead only logged.")
+}
+
+func main() {
+	cfg := config{}
+	cfg.registerFlags(flag.CommandLine)
+
+	// Parse CLI arguments.
+	if err := flagext.ParseFlagsWithoutArguments(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	if err := cfg.bucketConfig.Validate(); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	err := run(ctx, cfg)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func run(ctx context.Context, cfg config) error {
+	bucket, err := cfg.bucketConfig.ToBucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to create bucket")
+	}
+
+	blocks, err := getBlocks(ctx, cfg, bucket)
+	if err != nil {
+		return errors.Wrap(err, "failed to get blocks")
+	}
+
+	undeleteBlocks(ctx, bucket, blocks, cfg.dryRun)
+	return nil
+}
+
+type tenantFilter func(string) bool
+
+func newTenantFilter(cfg config) tenantFilter {
```

Comment:
considering this. If we wanted to bypass the tenant listing on `--blocks-from listing` with a `--include-tenants` it would have to be extended. might be better to try that out locally and then see if we wanted to add that to the `util` version.